File: Controllers/Api/TripsController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AutoMapper;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using TheWorld.Models;
using TheWorld.ViewModels;

namespace TheWorld.Controllers.Api
{
    [Route("api/trips")] //we centralise the route here instead of adding it nexto to each action method
    [Authorize]
    public class TripsController : Controller //although we won't return views but data as an API with MVC6
    {
        private ILogger<TripsController> _logger;
        private IWorldRepository _repository;

        public TripsController(IWorldRepository repository, ILogger<TripsController> logger) //this constructor will allow us to inject the IWorldRepository into our controller
                                                                                             //also we can use an ILogger for our specific class and create an object
        {
            _repository = repository;
            _logger = logger;  //as always, we generate a  global field (up) from the local variable we injected in the constructor
        }

        [HttpGet("")]
        public IActionResult Get()  //Retrieval of data. we could use JsonResult, however this will not allow me to catch errors.Instead of returning "Json" we will return just a helper "Ok"
        {
            //old code //if (true) return BadRequest("Bad things happened");

            //            /* return Ok(new Trip() { Name = "My Trip", UserName = "Grack" }); *///to test this i deactivated the url launcher inside properties

            try
            {
                var results = _repository.GetTripsByUsername(User.Identity.Name); //we store the user that claims the trips into result

                                                                                   //the variable stores a collection of the trip entity

                return Ok(Mapper.Map<IEnumerable<TripViewModel>>(results)); //as this is a get method, we are getting a collection //the mapper in startup also is configured to map from a collection of
                                                                            //entities, to a collection of viewmodels   
            }
            catch (Exception ex)
            {
                _logger.LogError($"Failed to get All Trips: {ex}");

                return BadRequest("Error occurred");
            }
        }

        [HttpPost("")]
        public async Task<IActionResult> Post([FromBody]TripViewModel theTrip) //! from body means: Please modelbind (map using the model) the data thats coming from the POST, and not from another URL(without this won't map it at all)
        {
            //modelbind,map it to this object by trying to match the json properties from the API!
            if (ModelState.IsValid)
            {
                // Save to the Database
                var newTrip = Mapper.Map<Trip>(theTrip);  //this maps the trip model into the "thetrip" object of type TripViewModel
                                                            //we need to configure this logic into startup

                newTrip.UserName = User.Identity.Name; //we assign the new created trip to the logged in user 

                _repository.AddTrip(newTrip);

                if (await _repository.SaveChangesAsync())
                {
                    return Created($"api/trips/{theTrip.Name}", Mapper.Map<TripViewModel>(newTrip));  //this returns and creates an object with an specified URI if the data is valid

                }
            }

            return BadRequest("Failed to save the trip"); //valud way to communicate the modelstate and any errors that are going down
        } 
    }
}
////we can call either an entity or a viewmodel. The viewmodel will hide the information to the client whereas the entity will show the parameters and the way our app is built

////when we use a viewmodel we need to convert this into an object (entity) that we can store into the database

////when we map create the object using an entity so we can store it to the database, we need to convert the viewmodel (if we use it)
////to entity (in this case from tripviewmodel to trip) and then create the object. However, we also need to convert it back to
////viewmodel, as otherwise we will be showing sensible contract shape to the client

//// viewmodel ---> map it so we can save it using model ---> viewmodel again with the newly created object 

File: wwwroot/js/tripsController.js
//tripsController.js
(function(){

    "use strict";
    //  Getting the existing the module
    //here the module is already created and we are referencing it to create a controller
    angular.module("app-trips")
        .controller("tripsController", tripsController); //create a controller under tripscontrollername with the below function
    
    function tripsController($http) {



        var vm = this; //more descriptive calling viewmodel than this all the time

        vm.trips = [];

        vm.newTrip = {}; //we will push the data using angular in our view to this viewmodel

        vm.errorMessage = ""; // non functional data. 

        vm.isBusy = true;

        $http.get("/api/trips") //the get will return a "promise" where we can get the results, errors or actions
                   .then(function (response) {
                       //success
                       angular.copy(response.data, vm.trips);

                   }, function (error) {
                       //failure
                       vm.errorMessage = "Failed to load data: " + error; //as it's JS the error will be shown if there is a failure
                   })
                    .finally(function () {
                     vm.isBusy = false;
                 });


        vm.addTrip = function () {
          
            vm.isBusy = true;
            vm.errorMessage = ""; //we restart the error message just in case we want to retry the post

            $http.post("/api/trips", vm.newTrip)//the post mthod takes and URI and the body (as postman). Body in this case is the newtrip 
                      .then(function (response) {
                             //success
                            vm.trips.push(response.data); //we push the data we take from the post method to our trips viewmodel
                            vm.newTrip = {}; //we clear the form if we succeed saving the new data into the server

                         }, function () {
                             //failure
                            vm.errorMessage = "Failed to save new trip";
                    }).finally(function () {
                            vm.isBusy = false; });
                     };
             }

    })();

//the controller is the responsible for the data, and we will use data binding to (ng-model , {{}}) to expose this data
//we use Cluenext sintax to add more actions to the function